src/org/exist/xslt/StylesheetUri.java

```diff
@@ -0,0 +1,189 @@
+      throws E, TransformerConfigurationException, IOException, PermissionDeniedException, SAXException {
+
+    if (uri.startsWith(XmldbURI.EMBEDDED_SERVER_URI_PREFIX)) {
+      final String docPath = uri.substring(XmldbURI.EMBEDDED_SERVER_URI_PREFIX.length());
+      DocumentImpl doc = null;
+      try {
+        doc = broker.getXMLResource(XmldbURI.create(docPath), LockMode.READ_LOCK);
+        if (doc == null) {
+          throw new IOException("XSL stylesheet not found: "+docPath);
+        }
+        if (templates == null || doc.getMetadata().getLastModified() > lastModified) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("compiling stylesheet " + doc.getURI());
+          }
+          templates = compileTemplates(broker, doc, errorListener);
+          lastModified = doc.getMetadata().getLastModified();
+        }
+      } finally {
+        if (doc != null) {
+          doc.getUpdateLock().release(LockMode.READ_LOCK);
+        }
+      }
+
+    } else {
+      final URL url = new URL(uri);
+      final URLConnection connection = url.openConnection();
+      long modified = connection.getLastModified();
+      if (templates == null || modified > lastModified || modified == 0) {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("compiling stylesheet " + url);
+        }
+        try (final InputStream is = connection.getInputStream()) {
+          templates = factory(broker.getBrokerPool(), errorListener).newTemplates(new StreamSource(is));
+        }
+      }
+      lastModified = modified;
+    }
+
+    return templates;
+  }
+
+  @Override
+  public <E extends Exception> TransformerHandler newTransformerHandler(DBBroker broker, XSLTErrorsListener<E> errorListener)
+      throws E, PermissionDeniedException, SAXException, TransformerConfigurationException, IOException {
+
+    TransformerHandler handler = cachedFactory(broker.getBrokerPool())
+        .newTransformerHandler(templates(broker, errorListener));
+
+    handler.getTransformer().setErrorListener(errorListener);
+
+    return handler;
+  }
+
+  private <E extends Exception> Templates compileTemplates(
+      DBBroker broker,
+      DocumentImpl stylesheet,
+      XSLTErrorsListener<E> errorListener)
+      throws E, TransformerConfigurationException, SAXException
+  {
+    //factory.setURIResolver(new EXistURIResolver(broker, stylesheet.getCollection().getURI().toString()));
+
+    final TemplatesHandler handler = factory(broker.getBrokerPool(), errorListener).newTemplatesHandler();
+
+    handler.startDocument();
+
+    final Serializer serializer = broker.newSerializer();
+    serializer.reset();
+    serializer.setSAXHandlers(handler, null);
+    serializer.toSAX(stylesheet);
+
+    handler.endDocument();
+
+    final Templates t = handler.getTemplates();
+
+    //check for errors
+    errorListener.checkForErrors();
+
+    return t;
+  }
+
+  private SAXTransformerFactory cachedFactory(BrokerPool db) {
```

Comment:
It seems to me this is used concurrently from `TemplatesFactory` so you will need synchronization on the cached `factory` and the mutable class members.